src/Proxy/ContainerInterfaceProxy.php

```diff
@@ -0,0 +1,260 @@
+            $timeStart = microtime(true);
+            $result = $this->container->has($id);
+        } catch (ContainerExceptionInterface $e) {
+            $this->repeatError($e);
+        } finally {
+            $this->log('has', [$id], $result, $timeStart);
+            return $result;
+        }
+    }
+
+    protected function getCurrentResultStatus(): string
+    {
+        return $this->currentError === null ? 'success' : 'failed';
+    }
+
+    protected function repeatError(object $error): void
+    {
+        $this->currentError = $error;
+        $errorClass = get_class($error);
+        throw new $errorClass($error->getMessage());
+    }
+
+    protected function resetCurrentError(): void
+    {
+        $this->currentError = null;
+    }
+
+    protected function log(string $method, array $arguments, $result, float $timeStart): void
+    {
+        $error = $this->currentError;
+        $this->processLogData($arguments, $result, $error);
+
+        if ($this->config->getCollector() !== null) {
+            $this->logToCollector($method, $arguments, $result, $error, $timeStart);
+        }
+
+        if ($this->config->getDispatcher() !== null) {
+            $this->logToEvent($method, $arguments, $result, $error, $timeStart);
+        }
+    }
+
+    private function processLogData(array &$arguments, &$result, ?object &$error): void
+    {
+        if (!($this->config->getLogLevel() & self::LOG_ARGUMENTS)) {
+            $arguments = null;
+        }
+
+        if (!($this->config->getLogLevel() & self::LOG_RESULT)) {
+            $result = null;
+        }
+
+        if (!($this->config->getLogLevel() & self::LOG_ERROR)) {
+            $error = null;
+        }
+    }
+
+    private function logToCollector(string $method, ?array $arguments, $result, ?object $error, float $timeStart): void
+    {
+        $this->config->getCollector()->collect(
+            ContainerInterface::class,
+            get_class($this->container),
+            $method,
+            $arguments,
+            $result,
+            $this->getCurrentResultStatus(),
+            $error,
+            $timeStart,
+            microtime(true),
+            );
+    }
+
+    private function logToEvent(string $method, ?array $arguments, $result, ?object $error, float $timeStart): void
+    {
+        $this->config->getDispatcher()->dispatch(new ProxyMethodCallEvent(
+            ContainerInterface::class,
+            get_class($this->container),
+            $method,
+            $arguments,
+            $result,
+            $this->getCurrentResultStatus(),
+            $error,
+            $timeStart,
+            microtime(true),
+            ));
+    }
+
+    private function isDecorated(string $service): bool
+    {
+        return isset($this->config->getDecoratedServices()[$service]) || in_array($service, $this->config->getDecoratedServices(), true);
+    }
+
+    private function getServiceProxy(string $service, object $instance): ?object
```

Comment:
Difficult to read. It is better to take it out in methods. For example:

```php
    private function getServiceProxy(string $service, object $instance): ?object
    {
        if (!$this->isDecorated($service)) {
            return null;
        }

        if ($this->config->hasDecoratedService($service) && is_callable($this->config->getDecoratedService($service))) {
            return $this->getServiceProxyFromCallable($service, $instance);
        }

        if ($this->config->hasDecoratedService($service) && $this->config->isArrayDecoratedService($service)) {
            if(!$this->config->hasFirstDecoratedService($service)) {
                return $this->getCommonMethodProxy($service, $instance, $this->config->getDecoratedService($service));
            }

            return $this->getServiceProxyFromArray($service, $instance);
        }

        if (interface_exists($service) && ($this->config->hasCollector() || $this->config->hasDispatcher())) {
            return $this->getCommonServiceProxy($service, $instance);
        }

        return null;
    }
```

ContainerProxyConfig:
```php

    public function hasDecoratedService(string $service): bool
    {
        return isset($this->decoratedServices[$service]);
    }

    public function isArrayDecoratedService(string $service): bool
    {
        return is_array($this->decoratedServices[$service]);
    }

    public function hasFirstDecoratedService(string $service): bool
    {
        return isset($this->decoratedServices[$service][0]);
    }

    public function getDecoratedService(string $service)
    {
        return $this->decoratedServices[$service];
    }

    public function hasDispatcher(): bool
    {
        return $this->dispatcher !== null;
    }

    public function hasCollector(): bool
    {
        return $this->collector !== null;
    }
```

Reply:
This is just a result after the auto-replace. It will be refactored.

Reply:
Done